fetch.js

```diff
@@ -9,34 +9,87 @@
 import fetch from 'isomorphic-unfetch';
 import { bookApiUrl } from './env';
 import type { RemoteData, Language, Book } from './types';
+import {
+  getAccessTokenFromLocalStorage,
+  setAnonToken,
+} from './lib/auth/token';
+
+let getTokenOnServer;
+
+if (!process.browser) {
+  getTokenOnServer = require('./server/lib/auth').getToken; // eslint-disable-line global-require
+}
+
+/**
+ * Get anonymous access token
+ */
+export async function fetchAnonToken(): Promise<{
+  access_token: string,
+  expires_in: number,
+}> {
+  if (!process.browser) {
+    const token = await getTokenOnServer();
+    return token;
+  }
+  const response = await fetch('/get_token');
+  if (response.ok) {
+    const token = await response.json();
+    return token;
+  }
+  throw new Error('Unable to get access token');
+}
 
 /*
 * Wrap fetch with some error handling and automatic json parsing
+* Also ensures we have a valid access token.
 */
-async function doFetch(url: string): Promise<RemoteData<any>> {
-  try {
-    const response = await fetch(url);
-
-    if (response.headers.get('Content-Type').includes('application/json')) {
-      const json = await response.json();
-      // Check if the response is in the 200-299 range
-      if (response.ok) {
-        return json;
-      }
+function fetchWithToken(
+  url: string,
+): (accessToken: ?string) => Promise<RemoteData<any>> {
+  return async (accessToken: ?string) => {
+    if (!process.browser && !accessToken) {
+      throw new Error(
+        'accessToken is a required parameter when calling fetch on the server',
+      );
     }
-    const err = new Error('Remote data error');
-    // $FlowFixMe Ignorore the flow error here. Should really extend the Error class, but that requires special Babel configuration. See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error#Custom_Error_Types
-    err.statusCode = response.status || 500;
-    throw err;
-  } catch (error) {
-    if (!error.statusCode) {
-      error.statusCode = 500;
+
+    let token = accessToken || getAccessTokenFromLocalStorage();
+
+    try {
+      // Automatically renew token on the client if it is expired
+      if (process.browser && accessToken == null) {
+        const fullToken = await fetchAnonToken();
+        setAnonToken(fullToken);
+        token = fullToken.access_token;
+      }
+
+      const response = await fetch(url, {
+        headers: {
+          Authorization: token ? `Bearer ${token}` : null,
```

Comment:
Is this a plausible situation, with `token === null`? What will happen then?

Reply:
You get rate limited as a completely anonymous user.